Add support for rudimentary conditionals in profiles

crass opened the pull request:
This was suggested in #2147.  The feature is designed to be extensible, though currently only `HAS_APPIMAGE` is supported.  The conditional profile line is of the form `?` + `<conditional>` + ` <normal profile line>`.  I've included in `whitelist-common.inc` the line `?HAS_APPIMAGE whitelist ${HOME}/.local/share/appimagekit`, which demonstrates usage of this feature and solves #2147.

It should be noted that conditionals cannot be used for `quiet` or `include` profile lines.  There is appropriate error messaging if the user tries this.

reinerh commented on src/firejail/profile.c:
```diff
@@ -113,12 +113,65 @@ void profile_add_ignore(const char *str) {
+int profile_check_conditional(char *ptr, int lineno, const char *fname) {
+	struct cond_t {
+		char *name;	// conditional name
+		size_t len;	// length of name
```
having to manually specify the length seems a bit fragile, when you anyway have only constant strings.
i would suggest to drop the len from the struct and get the length with strlen when needed.

crass replied:
If you look at `profile_check_line`, every 5th line has the problem of using literal numerals instead of strlen.  I don't think it would be an exaggeration to say that its pervasive.  However, don't like the error prone aspect either.  I've updated the pull request branch to get rid of the literal, but not the len field.

smitsohu commented:
Hmm, I tried to run this and I got corrupted memory when the sandbox was closed. Is that only me?
```
smitsohu@smitsohu:~$ exit
exit
*** Error in `firejail': free(): invalid pointer: 0x0000561ce3a5566e ***
======= Backtrace: =========
/lib/x86_64-linux-gnu/libc.so.6(+0x70bfb)[0x7fac8a9a0bfb]
/lib/x86_64-linux-gnu/libc.so.6(+0x76fc6)[0x7fac8a9a6fc6]
/lib/x86_64-linux-gnu/libc.so.6(+0x7780e)[0x7fac8a9a780e]
firejail(+0x674e)[0x561ce2ce874e]
/lib/x86_64-linux-gnu/libc.so.6(__libc_start_main+0xf1)[0x7fac8a9502e1]
firejail(+0xbafa)[0x561ce2cedafa]
======= Memory map: ========
561ce2ce2000-561ce2d3b000 r-xp 00000000 fe:02 1730702                    /usr/bin/firejail
561ce2f3b000-561ce2f3c000 r--p 00059000 fe:02 1730702                    /usr/bin/firejail
561ce2f3c000-561ce2f3d000 rw-p 0005a000 fe:02 1730702                    /usr/bin/firejail
561ce2f3d000-561ce303f000 rw-p 00000000 00:00 0 
561ce3a54000-561ce3a75000 rw-p 00000000 00:00 0                          [heap]
7fac84000000-7fac84021000 rw-p 00000000 00:00 0 
7fac84021000-7fac88000000 ---p 00000000 00:00 0 
7fac89ed0000-7fac89ee6000 r-xp 00000000 fe:02 786439                     /lib/x86_64-linux-gnu/libgcc_s.so.1
7fac89ee6000-7fac8a0e5000 ---p 00016000 fe:02 786439                     /lib/x86_64-linux-gnu/libgcc_s.so.1
7fac8a0e5000-7fac8a0e6000 r--p 00015000 fe:02 786439                     /lib/x86_64-linux-gnu/libgcc_s.so.1
7fac8a0e6000-7fac8a0e7000 rw-p 00016000 fe:02 786439                     /lib/x86_64-linux-gnu/libgcc_s.so.1
7fac8a0e8000-7fac8a0f2000 r-xp 00000000 fe:02 796926                     /lib/x86_64-linux-gnu/libnss_files-2.24.so
7fac8a0f2000-7fac8a2f2000 ---p 0000a000 fe:02 796926                     /lib/x86_64-linux-gnu/libnss_files-2.24.so
7fac8a2f2000-7fac8a2f3000 r--p 0000a000 fe:02 796926                     /lib/x86_64-linux-gnu/libnss_files-2.24.so
7fac8a2f3000-7fac8a2f4000 rw-p 0000b000 fe:02 796926                     /lib/x86_64-linux-gnu/libnss_files-2.24.so
7fac8a2f4000-7fac8a2fa000 rw-p 00000000 00:00 0 
7fac8a300000-7fac8a30b000 r-xp 00000000 fe:02 796928                     /lib/x86_64-linux-gnu/libnss_nis-2.24.so
7fac8a30b000-7fac8a50a000 ---p 0000b000 fe:02 796928                     /lib/x86_64-linux-gnu/libnss_nis-2.24.so
7fac8a50a000-7fac8a50b000 r--p 0000a000 fe:02 796928                     /lib/x86_64-linux-gnu/libnss_nis-2.24.so
7fac8a50b000-7fac8a50c000 rw-p 0000b000 fe:02 796928                     /lib/x86_64-linux-gnu/libnss_nis-2.24.so
7fac8a510000-7fac8a524000 r-xp 00000000 fe:02 794094                     /lib/x86_64-linux-gnu/libnsl-2.24.so
7fac8a524000-7fac8a724000 ---p 00014000 fe:02 794094                     /lib/x86_64-linux-gnu/libnsl-2.24.so
7fac8a724000-7fac8a725000 r--p 00014000 fe:02 794094                     /lib/x86_64-linux-gnu/libnsl-2.24.so
7fac8a725000-7fac8a726000 rw-p 00015000 fe:02 794094                     /lib/x86_64-linux-gnu/libnsl-2.24.so
7fac8a726000-7fac8a728000 rw-p 00000000 00:00 0 
7fac8a728000-7fac8a72f000 r-xp 00000000 fe:02 794096                     /lib/x86_64-linux-gnu/libnss_compat-2.24.so
7fac8a72f000-7fac8a92e000 ---p 00007000 fe:02 794096                     /lib/x86_64-linux-gnu/libnss_compat-2.24.so
7fac8a92e000-7fac8a92f000 r--p 00006000 fe:02 794096                     /lib/x86_64-linux-gnu/libnss_compat-2.24.so
7fac8a92f000-7fac8a930000 rw-p 00007000 fe:02 794096                     /lib/x86_64-linux-gnu/libnss_compat-2.24.so
7fac8a930000-7fac8aac5000 r-xp 00000000 fe:02 787869                     /lib/x86_64-linux-gnu/libc-2.24.so
7fac8aac5000-7fac8acc5000 ---p 00195000 fe:02 787869                     /lib/x86_64-linux-gnu/libc-2.24.so
7fac8acc5000-7fac8acc9000 r--p 00195000 fe:02 787869                     /lib/x86_64-linux-gnu/libc-2.24.so
7fac8acc9000-7fac8accb000 rw-p 00199000 fe:02 787869                     /lib/x86_64-linux-gnu/libc-2.24.so
7fac8accb000-7fac8accf000 rw-p 00000000 00:00 0 
7fac8acd0000-7fac8ace0000 r-xp 00000000 fe:02 786668                     /lib/x86_64-linux-gnu/libapparmor.so.1.4.0
7fac8ace0000-7fac8aedf000 ---p 00010000 fe:02 786668                     /lib/x86_64-linux-gnu/libapparmor.so.1.4.0
7fac8aedf000-7fac8aee0000 r--p 0000f000 fe:02 786668                     /lib/x86_64-linux-gnu/libapparmor.so.1.4.0
7fac8aee0000-7fac8aee1000 rw-p 00010000 fe:02 786668                     /lib/x86_64-linux-gnu/libapparmor.so.1.4.0
7fac8aee8000-7fac8af00000 r-xp 00000000 fe:02 796931                     /lib/x86_64-linux-gnu/libpthread-2.24.so
7fac8af00000-7fac8b0ff000 ---p 00018000 fe:02 796931                     /lib/x86_64-linux-gnu/libpthread-2.24.so
7fac8b0ff000-7fac8b100000 r--p 00017000 fe:02 796931                     /lib/x86_64-linux-gnu/libpthread-2.24.so
7fac8b100000-7fac8b101000 rw-p 00018000 fe:02 796931                     /lib/x86_64-linux-gnu/libpthread-2.24.so
7fac8b101000-7fac8b105000 rw-p 00000000 00:00 0 
7fac8b108000-7fac8b12b000 r-xp 00000000 fe:02 786509                     /lib/x86_64-linux-gnu/ld-2.24.so
7fac8b326000-7fac8b32b000 rw-p 00000000 00:00 0 
7fac8b32b000-7fac8b32c000 r--p 00023000 fe:02 786509                     /lib/x86_64-linux-gnu/ld-2.24.so
7fac8b32c000-7fac8b32d000 rw-p 00024000 fe:02 786509                     /lib/x86_64-linux-gnu/ld-2.24.so
7fac8b32d000-7fac8b32f000 rw-p 00000000 00:00 0 
7ffd0664e000-7ffd0666f000 rw-p 00000000 00:00 0                          [stack]
7ffd067f0000-7ffd067f3000 r--p 00000000 00:00 0                          [vvar]
7ffd067f3000-7ffd067f5000 r-xp 00000000 00:00 0                          [vdso]

```

crass commented:
@smitsohu thanks for testing!  I had noticed this too, but it went away and I thought it was related to the build process.  Anyway, it came back so I looked into it and its now fixed in the latest branch.  This issue had no adverse side effects because as you suspected the process was aborting at the very end of `main`.

smitsohu commented on src/firejail/profile.c:
```diff
@@ -113,12 +113,85 @@ void profile_add_ignore(const char *str) {
+		// if set, continue processing statement in caller
+		if (cond->value) {
+			// move ptr to start of profile line
+			ptr = strdup(ptr + cond->len + 1);
```
I think you are out-of-bounds of the old ptr if nothing follows after the conditional

crass replied:
Good catch!  This is a perfect example of why I want as much code as possible not running `setuid`...

crass commented:
@smitsohu I've refactored and changed the code while addressing the out of bounds pointer access.  I've changed the syntax to have a `:` between the conditional and the profile line.  I've also reworked the error handling.  I think its more clear now that there isn't the same issue.

crass commented:
@reinerh thanks for catching the missing `:` in the docs.  The branch has been updated.

crass commented:
Ok, rebased onto master and will merge as soon as the CI completes.
